feat: new CLI documentation (#4051)
webpack/webpack.js.org: src/content/api/cli.md

| `--prefetch`    | string          | Prefetch this request                                                                                          |
| `--json`        | boolean, string | Prints result as JSON or store it in a file                                                                    |
| `--mode`        | string          | Defines the mode to pass to webpack                                                                            |
| `--version`     | boolean         | Get current version                                                                                            |
| `--stats`       | boolean, string | It instructs webpack on how to treat the stats                                                                 |
| `--no-stats`    | boolean         | Disables stats output                                                                                          |
| `--analyze`     | boolean         | It invokes `webpack-bundle-analyzer` plugin to get bundle information                                          |

> 

Reviewer: I think better write aliases together with flags

Author: Yes

### Aliases

We've set certain aliases for commonly used flags to make it convenient to work with webpack.

| Shortcut | Replaces        |
| -------- | --------------- |
| -c       | `--config`      |
| -m       | `--merge`       |
| -o       | `--output-path` |
| -t       | `--target`      |
| -w       | `--watch`       |
| -h       | `--hot`         |
| -d       | `--devtool`     |
| -j       | `--json`        |
| -v       | `--version`     |

### Negated Flags

> | Shortcut   | Replaces                                                      |
> | ---------- | ------------------------------------------------------------- |
> | --no-color | Disabled any color on the console                             |
> | --no-hot   | Disabled hot reloading if you have it enabled via your config |

Reviewer 2: ```suggestion
| Flag   | Description                                                      |
| ---------- | ------------------------------------------------------------- |
| --no-color | Disables any color on the console                             |
| --no-hot   | Disables hot reloading if you have it enabled via your config |
```

| --no-stats | Disables any compilation stats emitted by webpack             |

### Core Flags

Starting CLI v4 and webpack v5, CLI imports the entire configuration schema from webpack core to alow tuning almost every property from the command line.

Here's the list of all the core flags supported by webpack v5 with CLI v4 - [link](https://github.com/webpack/webpack-cli/tree/next/packages/webpack-cli#webpack-5)

> 

Reviewer: Can we highlight it and dive small example?

Author: Done

Read the [installation guide](/guides/installation) if you don't already have webpack and CLI installed.

## Commands

webpack-cli offers a variety of commands to make working with webpack easy.

| Command   | Alias | Description                                            |
| --------- | ----- | ------------------------------------------------------ |
| `init`    | c     | Initialize a new webpack configuration                 |
| `migrate` | m     | Migrate a configuration to a new version               |
| `loader`  | l     | Scaffold a loader repository                           |
| `plugin`  | p     | Scaffold a plugin repository                           |
| `info`    | i     | Outputs information about your system and dependencies |
| `serve`   | s     | Run the webpack Dev Server                             |

## Usage

### With configuration file

```bash

> webpack [--config webpack.config.js]

Reviewer 3: Are these `[]`  representing defaults? Maybe add a note if So.

Author: already added a section about default config below

```

See [configuration](/configuration) for the options in the configuration file.

### Without configuration file

> ```sh
> webpack <entry> [<entry>] -o <output-path>
> ```

Reviewer 3: Does this mean?

```suggestion
```sh
webpack --entry [<entry-file>] -o <output-path>
```
```

Author: both are valid actually, you can pass entries without entry flag

Read the [installation guide](/guides/installation) if you don't already have webpack and CLI installed.

## Commands

webpack-cli offers a variety of commands to make working with webpack easy. By default webpack ships with

| Command   | Alias | Description                                            |
| --------- | ----- | ------------------------------------------------------ |
| `init`    | c     | Initialize a new webpack configuration                 |
| `migrate` | m     | Migrate a configuration to a new version               |
| `loader`  | l     | Scaffold a loader repository                           |
| `plugin`  | p     | Scaffold a plugin repository                           |
| `info`    | i     | Outputs information about your system and dependencies |
| `serve`   | s     | Run the webpack Dev Server                             |

## Flags

> webpack-cli offers a variety of commands to make working with webpack easy. By default webpack ships with

Reviewer 4: ```suggestion
webpack-cli offers a variety of commands to make working with webpack easy. By default webpack ships with the following flags:
```

| `--version, -v`     | boolean         | Get current version                                                                                            |
| `--stats`           | boolean, string | It instructs webpack on how to treat the stats                                                                 |
| `--no-stats`        | boolean         | Disables stats output                                                                                          |
| `--analyze`         | boolean         | It invokes `webpack-bundle-analyzer` plugin to get bundle information                                          |

### Negated Flags

| Flag       | Description                                                   |
| ---------- | ------------------------------------------------------------- |
| --no-color | Disabled any color on the console                             |
| --no-hot   | Disabled hot reloading if you have it enabled via your config |
| --no-stats | Disables any compilation stats emitted by webpack             |

### Core Flags

> Starting CLI v4 and webpack v5, CLI imports the entire configuration schema from webpack core to alow tuning almost every property from the command line.

Reviewer 4: ```suggestion
Starting CLI v4 and webpack v5, CLI imports the entire configuration schema from webpack core to allow tuning almost every configuration option from the command line.
```

```

```bash
webpack src/index.js -o dist
```

This will bundle your source code with entry as `index.js`, and the output bundle file will have a path of `dist`.

```bash

> 	| Asset     | Size    | Chunks      | Chunk Names |

Reviewer 5: Add current cli output here

	|-----------|---------|-------------|-------------|
	| main.js | 1.54 kB | 0 [emitted] | index       |
	[0] ./src/index.js 51 bytes {0} [built]
	[1] ./src/others.js 29 bytes {0} [built]
```

```bash
webpack ./src/index.js ./src.index2.js -o dist/
```

This will form the bundle with both the files as separate entry points.

```bash

> 	| Asset     | Size    | Chunks        | Chunk Names   |

Reviewer 5: Same here

> example - config file lookup will be in order of .webpack/webpack.config.development.js > webpack.config.development.js > webpack.config.js

```md
'webpack.config',
'webpack.config.dev',
'webpack.config.development',
'webpack.config.prod',
'webpack.config.production',
'.webpack/webpack.config',
'.webpack/webpack.config.none',
'.webpack/webpack.config.dev',
'.webpack/webpack.config.development',
'.webpack/webpack.config.prod',
'.webpack/webpack.config.production',
'.webpack/webpackfile',
```

If `mode` is supplied, say `production` then config looking order will be -

`.webpack/webpack.config.production.* > .webpack/webpack.config.prod.* > webpack.config.production.* > webpack.config.prod.* > webpack.config.*`

## Common Options

W> Note that Command Line Interface has a higher precedence for the arguments you use it with than your configuration file. For instance, if you pass [`--mode="production"`](/configuration/mode/#usage) to webpack CLI and your configuration file uses `development`, `production` will be used.

__List all of the commands and flags available on the cli__

```bash
webpack --help

> webpack -h

Reviewer 5: We don't support help alias
```suggestion
```

Reviewer 3: Not yet. But I think we should. 
Should I create a PR for the same?

Reviewer 5: > But I think we should.

In my opinion, I don't think we need a alias for help, it is small enough already and is standard around almost all command line tools. Also, it is already so short to type, instead we should add alias for bigger flag names that are commonly use. You may look for such flags and adding PR for them would be great.

Reviewer 3: Makes sense 👍. I forgot we already support both `--help` and `help`

Author: yep let's not, removing this

```

> __Show help for a single command or flag__

Reviewer 5: ```suggestion
__Show help for a single command__
```

Author: works for flag as well

```bash
webpack <command> --help

> webpack --<flag> --h

Reviewer 5: Help on flags don't work
```suggestion
```

Reviewer 3: should be `webpack --help <command>` or ``webpack --help <flag>``

Reviewer 5: Yes, flag on help works.

Author: works both ways

Reviewer 5: Nah, only flag on help works.

```

```bash
webpack ./src/index.js -o dist
```

This will bundle your source code with entry as `index.js`, and the output bundle file will have a path of `dist`.

```bash
Hash: 07a1c2d056e28ff1cab2
Version: webpack 4.44.2
Time: 174ms
Built at: 10/10/2020
  Asset       Size  Chunks             Chunk Names
main.js  952 bytes       0  [emitted]  main
Entrypoint main = main.js
[0] ./src/index.js 30 bytes {0} [built]

> [1] ./src/others.js 0 bytes {0} [built]

Reviewer 4: This needs an update

Author: updated

```

```bash
webpack ./src/index.js ./src/others2.js -o dist/
```

This will form the bundle with both the files as separate entry points.

```bash
Hash: fad168056241c7181505
Version: webpack 4.44.2
Time: 175ms
Built at: 10/10/2020
  Asset        Size  Chunks             Chunk Names
main.js  1010 bytes       0  [emitted]  main
Entrypoint main = main.js
[0] multi ./src/index.js ./src/others2.js 40 bytes {0} [built]
[1] ./src/index.js 30 bytes {0} [built]
[2] ./src/others.js 0 bytes {0} [built]
[3] ./src/others2.js 0 bytes {0} [built]
```

## Default Configurations

CLI will look for some default configurations in the path of your project, here are the config files picked up by CLI.

If no `mode` is supplied via flags or config then this is the lookup order in increasing order

> example - config file lookup will be in order of .webpack/webpack.config.development.js > webpack.config.development.js > webpack.config.js

```md
'webpack.config',
'webpack.config.dev',
'webpack.config.development',
'webpack.config.prod',

> 'webpack.config.production',

Reviewer 4: If no `--mode` is specified this is a bit weird. No mode defaults to production, so it should also default to choosing the production config.

Author: Ah right mode defaults to production, this needs a little update on the CLI side to default to production config, will take care of it.

```md
'webpack.config',
'webpack.config.dev',
'webpack.config.development',
'webpack.config.prod',
'webpack.config.production',
'.webpack/webpack.config',
'.webpack/webpack.config.none',
'.webpack/webpack.config.dev',
'.webpack/webpack.config.development',
'.webpack/webpack.config.prod',
'.webpack/webpack.config.production',
'.webpack/webpackfile',
```

If `mode` is supplied, say `production` then config looking order will be -

`.webpack/webpack.config.production.* > .webpack/webpack.config.prod.* > webpack.config.production.* > webpack.config.prod.* > webpack.config.*`

## Common Options

W> Note that Command Line Interface has a higher precedence for the arguments you use it with than your configuration file. For instance, if you pass [`--mode="production"`](/configuration/mode/#usage) to webpack CLI and your configuration file uses `development`, `production` will be used.

__List all of the commands and flags available on the cli__

```bash
webpack --help
webpack -h
```

__Show help for a single command or flag__

```bash
webpack <command> --help

> webpack --<flag> --help

Reviewer 3: ```suggestion
webpack --help --<flag> 
```

Author: both are valid right?

Reviewer 3: I dont think so. Can you once try Locally?

Author: ![image](https://user-images.githubusercontent.com/21009455/95653469-a0ce6900-0b16-11eb-9a7d-3317d348be78.png)
 yep valid

Reviewer 3: Interesting, I was trying with `--mode`. 

![Screenshot at 2020-10-10 16-37-43](https://user-images.githubusercontent.com/46647141/95653540-15a1a300-0b17-11eb-9321-3e58b94b643e.png)

Reviewer 3: It works for some flags only. The behavior of `webpack --flag --help` is inconsistent. Hence I think we should document `webpack --help --flag`. 

![Screenshot at 2020-10-11 08-11-21](https://user-images.githubusercontent.com/46647141/95669047-77a2ec80-0b99-11eb-84a4-ae889924f940.png)

Author: yeah lets remove it

```bash
webpack --json

> webpack --json stats.json

Reviewer 3: Maybe add a note here `If you want to store stats as json instead of printing it, you can use -`

Author: Yeah sounds good

```

In every other case, webpack prints out a set of stats showing bundle, chunk and timing details. Using this option, the output can be a JSON object. This response is accepted by webpack's [analyse tool](https://webpack.github.io/analyse/), or chrisbateman's [webpack-visualizer](https://chrisbateman.github.io/webpack-visualizer/), or th0r's [webpack-bundle-analyzer](https://github.com/webpack-contrib/webpack-bundle-analyzer). The analyse tool will take in the JSON and provide all the details of the build in graphical form.

## Environment Options

When the webpack configuration [exports a function](/configuration/configuration-types/#exporting-a-function), an "environment" may be passed to it.

```bash
webpack --env production    # sets env.production == true
```

> The `--env` argument accepts accepts multiple values:

Reviewer 2: ```suggestion
The `--env` argument accepts multiple values:
```

Author: nice fixed

| Invocation                     | Resulting environment       |
| ------------------------------ | --------------------------- |

> | `webpack --env prod`           | `{ prod: true`              |

Reviewer 3: ```suggestion
| `webpack --env prod`           | `{ prod: true }`              |
```

## Negated Flags

| Shortcut   | Replaces                                                      |
| ---------- | ------------------------------------------------------------- |
| --no-color | Disabled any color on the console                             |
| --no-hot   | Disabled hot reloading if you have it enabled via your config |
| --no-stats | Disables any compilation stats emitted by webpack             |

## Analyzing Bundle

You can also use `webpack-bundle-analyzer` to analyze your output bundles emitted by webpack. You can use `--analyze` flag to invoke it via CLI.

```sh
webpack --analyze
```

W> Make sure you have `webpack-bundle-analyzer` installed in your project else CLI will prompt you to install it.

## Aliases

We've set certain aliases for commonly used command and flags

| Shortcut | Replaces                                                                |
| -------- | ----------------------------------------------------------------------- |
| -d       | `--debug --devtool cheap-module-eval-source-map --output-pathinfo`      |

> | -p       | `--mode production`, see [building for production](/guides/production/) |

Reviewer 3: I think we don't have this in v4. Can you confirm?

Author: yeah shortcuts need to be updated, next commit

| --no-stats | Disables any compilation stats emitted by webpack             |

## Core Flags

Starting CLI v4 and webpack v5, CLI imports the entire configuration schema from webpack core to alow tuning almost every property from the command line.

Here's the list of all the core flags supported by webpack v5 with CLI v4 - [link](https://github.com/webpack/webpack-cli/tree/next/packages/webpack-cli#webpack-5)

## Analyzing Bundle

You can also use `webpack-bundle-analyzer` to analyze your output bundles emitted by webpack. You can use `--analyze` flag to invoke it via CLI.

```sh
webpack --analyze
```

W> Make sure you have `webpack-bundle-analyzer` installed in your project else CLI will prompt you to install it.

## Aliases

We've set certain aliases for commonly used flags to make it convenient to work with webpack.

| Shortcut | Replaces        |
| -------- | --------------- |
| -c       | `--config`      |
| -m       | `--merge`       |
| -o       | `--output-path` |
| -t       | `--target`      |
| -w       | `--watch`       |
| -h       | `--hot`         |
| -d       | `--devtool`     |
| -j       | `--json`        |
| -v       | `--version`     |

> | -v       | `--version`     |

Reviewer 5: ```suggestion
```

Author: done
